File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodemanager/NodeManager.java
package org.opcfoundation.webserver.addressspace.nodemanager;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.*;
import org.eclipse.milo.opcua.stack.core.types.enumerated.BrowseDirection;
import org.eclipse.milo.opcua.stack.core.types.enumerated.BrowseResultMask;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.enumerated.TimestampsToReturn;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseResult;
import org.eclipse.milo.opcua.stack.core.types.structured.ReferenceDescription;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaNode;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.addressspace.nodes.UaReference;
import org.opcfoundation.webserver.addressspace.nodes.UaVariable;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;
import org.opcfoundation.webserver.types.common.UaBrowseContinuationPoint;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public abstract class NodeManager implements NodeManagerBase {

    private final int nsIndex;
    private final String namespaceUri;
    private final HashMap<NodeId, UaNode> nodes;

    public NodeManager(String namespaceUri)
    {
        this.nsIndex = NodeManagerList.nodeManagerList.getNewNsIndex();
        this.namespaceUri = namespaceUri;
        this.nodes = new HashMap<>();
    }

    @Override
    public int nsIndex() {
        return nsIndex;
    }

    @Override
    public String namespaceUri() {
        return namespaceUri;
    }

    @Override
    public @Nullable UaNode getNode(NodeId nodeId)
    {
        if (nsIndex() == nodeId.getNamespaceIndex().intValue())
        {
            return nodes.get(nodeId);
        } else {
            NodeManagerBase nodeManager = NodeManagerList.nodeManagerList.getNodeManager(nodeId.getNamespaceIndex().intValue());
            if (null == nodeManager) return null;
            return nodeManager.getNode(nodeId);
        }
    }

    public void addNode(UaNode node) throws UaRuntimeException
    {
        if (nodes.containsKey(node.nodeId())) throw new UaRuntimeException(StatusCodes.Bad_NodeIdExists);
        if (node.nodeId().getNamespaceIndex().intValue() != nsIndex()) throw new UaRuntimeException(StatusCodes.Bad_NodeIdInvalid);
        nodes.put(node.nodeId(), node);
    }

    public @Nullable List<UaReference> browseNode(
            NodeId nodeId,
            BrowseDirection browseDirection,
            NodeId referenceTypeId,
            Boolean includeSubtypes,
            int nodeClassMask)
    {
        UaNode nodeToBrowse = getNode(nodeId);
        if (null == nodeToBrowse) return null;

        List<UaReference> references = nodeToBrowse.getReferences(browseDirection);

        ArrayList<UaReference> refsToReturn = new ArrayList<>();

        for (UaReference item : references)
        {
            if ((item.linkedNode().nodeClass().getValue() & nodeClassMask) == 0) continue;

            if (includeSubtypes)
            {
                if (!item.reference().isSubtypeOf(referenceTypeId)) continue;
            } else {
                if (!item.reference().nodeId().equals(referenceTypeId)) continue;
            }

            refsToReturn.add(item);
        }

        return refsToReturn;
    }

    public BrowseResult browse(
            BrowseDescription browseDescription,
            UaBrowseAdditionalInfo additionalInfo)
    {
        List<UaReference> referencesOfNode = browseNode(
                browseDescription.getNodeId(),
                browseDescription.getBrowseDirection(),
                browseDescription.getReferenceTypeId(),
                browseDescription.getIncludeSubtypes(),
                browseDescription.getNodeClassMask().intValue());

        if (null == referencesOfNode) return new BrowseResult(StatusCode.of(StatusCodes.Bad_NodeIdUnknown), ByteString.NULL_VALUE, null);

        // Determinate the range
        boolean isCPRequired = false;
        int fromIndex = additionalInfo.getReferenceOffset();
        int toIndex = referencesOfNode.size();

        if (referencesOfNode.size() <= fromIndex) return new BrowseResult(StatusCode.of(StatusCodes.Good), ByteString.NULL_VALUE, null);

        if (0 < additionalInfo.getMaxReferencesPerNode())
        {
            toIndex = fromIndex + additionalInfo.getMaxReferencesPerNode();
            if (toIndex >= referencesOfNode.size())
            {
                toIndex = referencesOfNode.size();
            } else {
                isCPRequired = true;
            }
        }

        List<UaReference> referencesToReturn = referencesOfNode.subList(fromIndex, toIndex);
        ArrayList<ReferenceDescription> references = new ArrayList<>();

        for (UaReference item: referencesToReturn)
        {
            NodeId referenceTypeId = NodeId.NULL_VALUE;
            QualifiedName browseName = QualifiedName.NULL_VALUE;
            LocalizedText displayName = LocalizedText.NULL_VALUE;
            NodeClass nodeClass = NodeClass.Unspecified;
            NodeId typeDefinition = NodeId.NULL_VALUE;

            if (0 != (browseDescription.getResultMask().intValue() & BrowseResultMask.ReferenceTypeId.getValue()))
            {
                referenceTypeId = item.reference().nodeId();
            }

            if (0 != (browseDescription.getResultMask().intValue() & BrowseResultMask.BrowseName.getValue()))
            {
                browseName = new QualifiedName(0,item.linkedNode().browseName());
            }

            if (0 != (browseDescription.getResultMask().intValue() & BrowseResultMask.DisplayName.getValue()))
            {
                displayName =  item.linkedNode().displayName();
            }

            if (0 != (browseDescription.getResultMask().intValue() & BrowseResultMask.NodeClass.getValue()))
            {
                nodeClass = item.linkedNode().nodeClass();
            }

            if (0 != (browseDescription.getResultMask().intValue() & BrowseResultMask.TypeDefinition.getValue()))
            {
                if (item.linkedNode().nodeClass() == NodeClass.Object)
                {
                    typeDefinition = ((UaObject)item.linkedNode()).typeDefinition().nodeId();
                }

                if (item.linkedNode().nodeClass() == NodeClass.Variable)
                {
                    typeDefinition = ((UaVariable)item.linkedNode()).typeDefinition().nodeId();
                }
            }

            references.add(new ReferenceDescription(
                    referenceTypeId,
                    item.isForward(),
                    item.linkedNode().nodeId().expanded(),
                    browseName,
                    displayName,
                    nodeClass,
                    typeDefinition.expanded()));
        }

        ByteString continuationPoint = ByteString.NULL_VALUE;
        if (isCPRequired)
        {
            UaBrowseContinuationPoint cp = new UaBrowseContinuationPoint(browseDescription, new UaBrowseAdditionalInfo(additionalInfo.getMaxReferencesPerNode(), toIndex, 0));
            continuationPoint = cp.toByteString();
        }

        return new BrowseResult(StatusCode.of(StatusCodes.Good), continuationPoint, references.toArray(new ReferenceDescription[0]));
    }

    public DataValue read(
            NodeId nodeId,
            int attributeId,
            DateTime now,
            TimestampsToReturn timestampsToReturn)
    {
        UaNode nodeToRead = getNode(nodeId);

        Variant value = Variant.NULL_VALUE;
        StatusCode statusCode = StatusCode.GOOD;
        DateTime sourceTimestamp = null;
        DateTime serverTimestamp = null;

        if (null != nodeToRead)
        {
            value = nodeToRead.getAttribute(attributeId);
            if (value.isNull() && AttributeId.Value.id() != attributeId)
            {
                statusCode =  StatusCode.of(StatusCodes.Bad_AttributeIdInvalid);
            }
        } else {
            statusCode = StatusCode.of(StatusCodes.Bad_NodeIdUnknown);
        }

        if (statusCode.isGood() && AttributeId.Value.id() == attributeId)
        {
            if (TimestampsToReturn.Both == timestampsToReturn ||
                    TimestampsToReturn.Source == timestampsToReturn)
            {
                sourceTimestamp = now;
            }

            if (TimestampsToReturn.Both == timestampsToReturn ||
                    TimestampsToReturn.Server == timestampsToReturn)
            {
                serverTimestamp = now;
            }
        }

        return new DataValue(value, statusCode, sourceTimestamp, serverTimestamp);
    }
}

File: opcua-webservice-node/source/src/server/addressspace/nodemanager/NodeManager.ts
import { Attributes, BrowseDirection, NodeClass, StatusCodes, TimestampsToReturn } from "opcua-webapi";
import {
    isBrowseMaskRequired,
    makeUaStatusCode,
    UaBrowseDescription,
    UaBrowseMask,
    UaBrowseResult,
    UaDataValue,
    UaError,
    UaExpandedNodeId,
    UaLocalizedText,
    UaNodeId,
    UaReferenceDescription,
    UaVariant,
} from "opcua-webapi-ts";
import { UaBrowseAdditionalInfo, UaBrowseContinuationPoint } from "../../types";
import { UaNode, UaObject, UaReference, UaVariable } from "../nodes";
import { NodeManagerBase } from "./NodeManagerBase";
import { NodeManagerList } from "./NodeManagerList";

export abstract class NodeManager extends NodeManagerBase {
    private readonly _nsIndex: number;
    private readonly _namespaceUri: string;
    private readonly _nodes: Map<string, UaNode>;

    constructor(namespaceUri: string) {
        super();
        this._nsIndex = NodeManagerList.nodeManagerList.getNewNsIndex();
        this._namespaceUri = namespaceUri;
        this._nodes = new Map<string, UaNode>();        
    }   

    nsIndex(): number {
        return this._nsIndex;
    }

    namespaceUri(): string {
        return this._namespaceUri;
    }

    getNode(nodeId: UaNodeId): UaNode | null {
        if (this.nsIndex() === nodeId.nsIndex) {
            return this._nodes.get(nodeId.toString()) ?? null;
        }

        const nodeManager = NodeManagerList.nodeManagerList.getNodeManager(nodeId.nsIndex);
        if (nodeManager === null) {
            return null;
        }

        return nodeManager.getNode(nodeId);
    }

    addNode(node: UaNode): void {
        if (this._nodes.has(node.nodeId.toString())) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdExists));
        }

        if (node.nodeId.nsIndex !== this.nsIndex()) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdInvalid));
        }

        this._nodes.set(node.nodeId.toString(), node);
    }

    browseNode(
        nodeId: UaNodeId,
        browseDirection: number,
        referenceTypeId: UaNodeId,
        includeSubtypes: boolean,
        nodeClassMask: number): Array<UaReference> | null {
        const nodeToBrowse = this.getNode(nodeId);
        if (nodeToBrowse === null) return null;
  
        const references = nodeToBrowse.getReferences(browseDirection as BrowseDirection);
        const refsToReturn: Array<UaReference> = [];

        for (const item of references) {
            if (((item.linkedNode.nodeClass as number) & nodeClassMask) === 0) {
                continue;
            }

            if (includeSubtypes) {
                if (!item.reference.isSubtypeOf(referenceTypeId)) {
                    continue;
                }
            } else if (!item.reference.nodeId.equal(referenceTypeId)) {
                continue;
            }

            refsToReturn.push(item);
        }

        return refsToReturn;
    }

    browse(
        browseDescription: UaBrowseDescription,
        additionalInfo: UaBrowseAdditionalInfo,
    ): UaBrowseResult {
        const referencesOfNode = this.browseNode(
            browseDescription.nodeId,
            browseDescription.browseDirection,
            browseDescription.referenceTypeId,
            browseDescription.includeSubtypes,
            browseDescription.nodeClassMask,
        );

        if (referencesOfNode === null) {
            return new UaBrowseResult([], null, makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
        }

        let isContinuationPointRequired = false;
        const fromIndex = additionalInfo.referenceOffset;
        let toIndex = referencesOfNode.length;

        if (referencesOfNode.length <= fromIndex) {
            return new UaBrowseResult([], null, makeUaStatusCode(StatusCodes.Good));
        }

        if (additionalInfo.maxReferencesPerNode > 0) {
            toIndex = fromIndex + additionalInfo.maxReferencesPerNode;
            if (toIndex >= referencesOfNode.length) {
                toIndex = referencesOfNode.length;
            } else {
                isContinuationPointRequired = true;
            }
        }

        const referencesToReturn = referencesOfNode.slice(fromIndex, toIndex);
        const references: Array<UaReferenceDescription> = [];

        for (const item of referencesToReturn) {
            const linkedNode = item.linkedNode;
            const resultMask = browseDescription.resultMask;

            let refTypeId: UaNodeId | undefined;
            let browseName: string | undefined;
            let displayName : UaLocalizedText | undefined;
            let nodeClass: number | undefined;
            let typeDefinition: UaExpandedNodeId | undefined;            

            if (isBrowseMaskRequired(resultMask , UaBrowseMask.ReferenceTypeId)) {
                refTypeId = item.reference.nodeId;
            }

            if (isBrowseMaskRequired(resultMask , UaBrowseMask.BrowseName)) {
                browseName = linkedNode.browseName;
            }

            if (isBrowseMaskRequired(resultMask , UaBrowseMask.DisplayName)) {
                displayName = linkedNode.displayName;
            }

            if (isBrowseMaskRequired(resultMask , UaBrowseMask.NodeClass)) {
                nodeClass = linkedNode.nodeClass;
            }

            if (isBrowseMaskRequired(resultMask , UaBrowseMask.TypeDefinition)) {
                if (linkedNode.nodeClass === NodeClass.Object) {
                    typeDefinition = UaExpandedNodeId.from((linkedNode as UaObject).typeDefinition.nodeId);
                } else if (linkedNode.nodeClass === NodeClass.Variable) {
                    typeDefinition = UaExpandedNodeId.from((linkedNode as UaVariable).typeDefinition.nodeId);
                }
            }

            references.push(
                new UaReferenceDescription(
                    UaExpandedNodeId.from(linkedNode.nodeId),
                    nodeClass,
                    browseName,
                    displayName,
                    refTypeId,
                    item.isForward,
                    typeDefinition,
                ),
            );
        }

        let continuationPoint: string | null = null;
        if (isContinuationPointRequired) {
            const cp = new UaBrowseContinuationPoint(
                browseDescription,
                new UaBrowseAdditionalInfo(additionalInfo.maxReferencesPerNode, toIndex, 0),
            );
            continuationPoint = cp.toByteString();
        }

        return new UaBrowseResult(references, continuationPoint, makeUaStatusCode(StatusCodes.Good));
    }

    read(
        nodeId: UaNodeId,
        attributeId: number,
        now: Date,
        timestampsToReturn: TimestampsToReturn,
    ): UaDataValue {
        const nodeToRead = this.getNode(nodeId);

        let value = UaVariant.null();
        let statusCode = makeUaStatusCode(StatusCodes.Good);
        let sourceTimestamp: Date | undefined;
        let serverTimestamp: Date | undefined;

        if (nodeToRead !== null) {
            value = nodeToRead.getAttribute(attributeId);
            if (value.isNull() && Attributes.Value !== attributeId) {
                statusCode = makeUaStatusCode(StatusCodes.BadAttributeIdInvalid);
            }
        } else {
            statusCode = makeUaStatusCode(StatusCodes.BadNodeIdUnknown);
        }

        if (statusCode.isGood() && Attributes.Value === attributeId) {
            if (timestampsToReturn === TimestampsToReturn.Both ||
                timestampsToReturn === TimestampsToReturn.Source) {
                sourceTimestamp = now;
            }

            if (timestampsToReturn === TimestampsToReturn.Both ||
                timestampsToReturn === TimestampsToReturn.Server) {
                serverTimestamp = now;
            }
        }

        return new UaDataValue(value, statusCode, sourceTimestamp, serverTimestamp);
    }
}
